Azure Monitor: Updated `grafana_ds_azuremonitor_dashboard_loaded` event, replaced array of queries for stats

yaelleC opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Tweaking our `grafana_ds_azuremonitor_dashboard_loaded` event to:
* remove user_id since some include the user email (privacy) and a better option (anonymous_id) is available out of the box in Rudderstack
* replace the huge array of queries prop for a much neater option: a few stats props (one per queryType/hidden combinaison = 6)

This PR also extracts the event tracking to a new file for documentation purposes

**Special notes for your reviewer**:

yaelleC committed:
Replaced array of queries for stats

yaelleC committed:
Added ds_version prop

yaelleC committed:
Added ds_version to tests

yaelleC commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/module.ts:
```diff
@@ -16,16 +16,22 @@ getAppEvents().subscribe<DashboardLoadedEvent<AzureMonitorQuery>>(
+        (query.queryType?.toLowerCase().replaceAll(' ', '_') ?? 'unknown') +
+        '_queries_' +
+        (query.hide ? 'hidden' : 'executed');
+      stats[statName] = ~~stats[statName] + 1;
```
This is a neat way, and will scale if we add more query types, but that means the event structure will change a few times for RS/BQ (first event probably won't have all types of query hidden and executed, so columns will be added by subsequent events triggered).

Maybe it's best I initialise all 6 props (Monitor, Logs, ARG for hidden and executed) to 0 to avoid the schema changes?

sunker replied:
Yeah I think declaring the props will be easier to understand.

asimpson replied:
Ditto.

yaelleC committed:
Extracted event name to tracking file

yaelleC committed:
Extracted event to tracking file

yaelleC commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/module.ts:
```diff
@@ -16,16 +17,20 @@ getAppEvents().subscribe<DashboardLoadedEvent<AzureMonitorQuery>>(
-      reportInteraction('grafana_ds_azuremonitor_dashboard_loaded', {
+      TrackAzureMonitorDashboardLoaded({
         grafana_version: grafanaVersion,
+        ds_version: pluginJson.info.version,
```
any preference between `ds_version` and `plugin_version`?

yaelleC committed:
Removed ds_version - useless for core plugin

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/tracking.ts:
```diff
@@ -0,0 +1,36 @@
+ * - v9.1.0 : list of queries logged
+ * - v9.1.2 : list removed in favour of stats, user_id removed
+ */
+export const TrackAzureMonitorDashboardLoaded = (props: AzureMonitorDashboardLoadedProps) => {
```
Named exports should be camelCased

sunker reviewed (commented):
Regarding hide/executed in the tracking event. It's not certain that a query will be executed just because it's not hidden. It may also be [filtered out](https://github.com/grafana/grafana/blob/update_az_adoption_event/public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts#L60) because it's not complete. You _can_ run the query through the `filterQuery` check before aggregating the tracking data. But keep in mind that it will also not be completely accurate since at this stage of the app life cycle, none of the plugin specific [query migration](https://github.com/grafana/grafana/blob/update_az_adoption_event/public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts#L18) have been executed yet.

yaelleC committed:
Addressed comments

asimpson reviewed (approved):
LGTM, nice work on this! :trophy:

yaelleC committed:
Added note to event documentation

yaelleC commented:
> Regarding hide/executed in the tracking event. It's not certain that a query will be executed just because it's not hidden. It may also be [filtered out](https://github.com/grafana/grafana/blob/update_az_adoption_event/public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts#L60) because it's not complete. You _can_ run the query through the `filterQuery` check before aggregating the tracking data. But keep in mind that it will also not be completely accurate since at this stage of the app life cycle, none of the plugin specific [query migration](https://github.com/grafana/grafana/blob/update_az_adoption_event/public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts#L18) have been executed yet.

For visibility: 

* I renamed the `hidden` / `executed` to avoid confusion

* Running `filterQuery` and migrations seems like a good improvement, unfortunately, I was not able to do so from `module.ts` (we can't really import the datasourceSrv from there) and I would like to avoid overcrowding `initDashboard` for now. 
As it is, the Azure Monitor event should not be impacted by migrations as it only looks at query types. The `filterQuery` would be useful in filtering out invalid/empty queries, but again even without it the data still provides us useful insights.
I added a note in the event documentation to make it clear that the queries are pre-migration and filterQuery.

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/module.ts:
```diff
@@ -16,16 +17,41 @@ getAppEvents().subscribe<DashboardLoadedEvent<AzureMonitorQuery>>(
+        azure_monitor_queries: stats[AzureQueryType.AzureMonitor].visible,
+        azure_log_analytics_queries: stats[AzureQueryType.LogAnalytics].visible,
+        azure_resource_graph_queries: stats[AzureQueryType.AzureResourceGraph].visible,
+        azure_monitor_queries_hidden: stats[AzureQueryType.AzureMonitor].hidden,
```
another nit: the distinction between visible and hidden seems a bit arbitrary. I know this is some simplified version of the query filtering but if we are just interested on visible queries, I would just record that.

andresmgot reviewed (commented):
Just got to this PR, I have some minor feedback, let me know what you think!
